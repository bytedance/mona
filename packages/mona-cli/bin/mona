#! /usr/bin/env node

const mona = require('../dist').default;
mona();
