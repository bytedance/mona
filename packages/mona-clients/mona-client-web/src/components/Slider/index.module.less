.slider {
  margin: 10px 18px;
  padding: 0;
  display: block;
  & * {
    -webkit-tap-highlight-color: rgba(0,0,0,0);
  }

  .wrapper {
    display: -webkit-flex;
    display: flex;
    -webkit-align-items: center;
    align-items: center;
    min-height: 16px;
  }

  .tapArea {
    -webkit-flex: 1;
    flex: 1;
    padding: 8px 0;
  }

  .handleWrapper {
    position: relative;
    height: 2px;
    border-radius: 5px;
    background-color: #e9e9e9;
    cursor: pointer;
    transition: background-color 0.3s ease;
    -webkit-tap-highlight-color: transparent;
  }

  .sliderTrack {
    height: 100%;
    border-radius: 6px;
    background-color: #1aad19;
    transition: background-color 0.3s ease;
  }

  .sliderHandle, .sliderThumb {
    position: absolute;
    left: 50%;
    top: 50%;
    cursor: pointer;
    border-radius: 50%;
    transition: border-color 0.3s ease;
  }

  .sliderHandle {
    width: 28px;
    height: 28px;
    margin-top: -14px;
    margin-left: -14px;
    background-color: transparent;
    z-index: 3;
    border: 0 solid rgba(0,0,0,0.04);
  }

  .sliderThumb {
    z-index: 2;
    box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2);
  }

  .sliderStep {
    position: absolute;
    width: 100%;
    height: 2px;
    background: transparent;
    z-index: 1;
  }

  .sliderValue {
    color: #888;
    font-size: 14px;
    margin-left: 1em;
  }

  .sliderDisabled {
    .sliderTrack {
      background-color: #ccc;
    }

    .sliderThumb {
      background-color: #FFF;
      border-color: #ccc;
    }
  }
}