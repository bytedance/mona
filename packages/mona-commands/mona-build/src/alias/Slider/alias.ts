import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { SliderAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  min: propAliasMap.min,
  max: propAliasMap.max,
  step: propAliasMap.step,
  disabled: propAliasMap.disabled,
  value: propAliasMap.value,
  color: propAliasMap.color,
  selectedColor: propAliasMap.selectedColor,
  activeColor: propAliasMap.activeColor,
  backgroundColor: propAliasMap.backgroundColor,
  blockColor: propAliasMap.blockColor,
  showValue: propAliasMap.showValue,
  onChange: propAliasMap.onChange,
  onChanging: propAliasMap.onChanging,
  blockSize: propAliasMap.blockSize,
};

export default alias;
