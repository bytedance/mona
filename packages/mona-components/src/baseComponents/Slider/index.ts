import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface SliderProps extends BaseProps {
  min?: number;
  max?: number;
  step?: number;
  disabled?: boolean;
  value?: number;
  color?: string;
  selectedColor?: string;
  activeColor?: string;
  backgroundColor?: string;
  blockSize?: number;
  blockColor?: string;
  showValue?: boolean;
  onChange?: EventHandler;
  onChanging?: EventHandler;
}

export const Slider: React.ComponentType<SliderProps> = createComponent<SliderProps>('slider')

