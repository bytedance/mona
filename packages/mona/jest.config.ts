import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
