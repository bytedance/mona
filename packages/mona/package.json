{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
