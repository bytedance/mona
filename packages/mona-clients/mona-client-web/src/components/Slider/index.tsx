import React, { useCallback, useEffect, useRef, useState } from 'react';
import { SliderProps } from '@bytedance/mona';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { useFormContext } from '../Form/hooks';

function formatValue(value: number, min: number, max: number, step: number) {
  let targetValue = value < min ? min : value > max ? max : value;
  const remainder = (targetValue - min) % step;
  if (remainder >= step / 2) {
    targetValue = targetValue - remainder + step;
  } else {
    targetValue = targetValue - remainder;
  }

  const decimalLength = (String(step).split('.')[1] || '').length;

  return Number(Number(targetValue).toFixed(decimalLength));
}

function formatSize(value: number) {
  return value < 12 ? 12 : value > 28 ? 28 : value;
}

function getPercent(value: number, min: number, max: number) {
  return (value - min) / (max - min) * 100;
}

const Slider: React.FC<SliderProps> = (props) => {
  const {
    children,
    min = 0,
    max = 100,
    step = 1,
    disabled = false,
    value = 0,
    color = '#e9e9e9',
    selectedColor = '#1aad19',
    activeColor = '#1aad19',
    backgroundColor = '#e9e9e9',
    blockSize = 28,
    name = '',
    blockColor = '#ffffff',
    showValue = false,
    onChange,
    onChanging,
    ...restProps
  } = props;
  const [v, setV] = useState(formatValue(value, min, max, step));
  const thumbRef = useRef<HTMLDivElement | null>(null);

   useEffect(() => {
    setV(formatValue(value, min, max, step));
  }, [value, min, max, step])

  const reset = useCallback(() => setV(min), []);
  useFormContext(name, value, reset);

  const size = formatSize(blockSize);

  const { handleClassName, ...handlerProps } = useHandlers(restProps);

  const handleClick = (e: React.MouseEvent) => {
    const target = e.currentTarget;
    const { left } = target.getBoundingClientRect();
    const width = target.clientWidth;
    const gapWidth = e.clientX - left;
    const ratio = gapWidth / width;
    const newValue = (max - min) * ratio + min;
    setV(formatValue(newValue, min, max, step));
  }

  const percent = getPercent(v, min, max);

  const validTouchDistanceRef = useRef(30);
  const wrapperRef = useRef<HTMLDivElement | null>(null);
  useEffect(() => {
    if (wrapperRef.current) {
      const width = wrapperRef.current.offsetWidth;
      const result = width / (max - min) * step;
      const d = validTouchDistanceRef.current;
      validTouchDistanceRef.current = d < result ? d : result;
    }
  }, [wrapperRef.current, max, min, step])

  const startThumbRef = useRef<{ x: number } | null>(null);
  const handleTouchStart = (e: React.TouchEvent) => {
    if (e.target === thumbRef.current)  {
      const { changedTouches } = e;
      startThumbRef.current = {
        x: changedTouches[0].clientX,
      }
    }
  }

  const handleTouchMove = (e: React.TouchEvent) => {
    // console.log('move');
    if (!startThumbRef.current) {
      return;
    }

    const { changedTouches } = e;
    const moveX = changedTouches[0].pageX;
    const { x } = startThumbRef.current;

    const dx = moveX - x;
    const abs = Math.abs(dx);
    const dv = validTouchDistanceRef.current;
    if (abs >= dv) {
      startThumbRef.current = {
        x: moveX
      }
      const pos = dx / abs;
      const newValue = v + pos * step * (abs / dv);
      setV(formatValue(newValue, min, max, step));
    }
  }

  const handleTouchEnd = () => {
    startThumbRef.current = null;
  }
  
  return (
    <div className={handleClassName(styles.slider)} {...handlerProps}>
      <div className={styles.wrapper}>
        <div className={styles.tapArea}>
          <div className={styles.handleWrapper} ref={wrapperRef} onTouchStart={handleTouchStart} onTouchMove={handleTouchMove} onTouchEnd={handleTouchEnd} onClick={handleClick} style={{ backgroundColor: backgroundColor || color }}>
            <div className={styles.sliderHandle} ref={thumbRef} style={{ left: `${percent}%` }}></div>
            <div className={styles.sliderThumb} style={{ width: size, height: size, marginLeft: -size/2, marginTop: -size/2, left: `${percent}%`, backgroundColor: blockColor }}></div>
            <div className={styles.sliderTrack} style={{ width: `${percent}%`, backgroundColor: activeColor || selectedColor }}></div>
            <div className={styles.sliderStep}></div>
          </div>
        </div>
        { showValue ? <span className={styles.sliderValue}>{v}</span> : null }
      </div>
    </div>
  )
}

export default Slider;
