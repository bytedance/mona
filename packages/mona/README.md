# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
