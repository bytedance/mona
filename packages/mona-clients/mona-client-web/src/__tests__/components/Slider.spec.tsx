import React from 'react';
import { mount } from 'enzyme';
import Slider from '../../components/Slider';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { SliderProps } from '@bytedance/mona';

describe('web component: Slider', () => {
  mountTest(Slider)
  handlerTest({ Component: Slider, firstTag: 'div' });

  it('should render correctly', () => {
    expect(mount(<Slider />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: SliderProps = {
      name: 'a',
      min: 10,
      max: 100,
      step: 10,
      disabled: false,
      value: 20,
      color: 'red',
      selectedColor: 'blue',
      activeColor: 'yellow',
      backgroundColor: 'grey',
      blockSize: 10,
      blockColor: 'skublue',
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<Slider {...props}>hello</Slider>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
