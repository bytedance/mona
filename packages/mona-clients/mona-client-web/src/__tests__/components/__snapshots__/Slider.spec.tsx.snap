// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: Slider should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="slider class-name  "
    style="color: red;"
  >
    <div
      class="wrapper"
    >
      <div
        class="tapArea"
      >
        <div
          class="handleWrapper"
          style="background-color: grey;"
        >
          <div
            class="sliderHandle"
            style="left: 11.11111111111111%;"
          />
          <div
            class="sliderThumb"
            style="width: 12px; height: 12px; margin-left: -6px; margin-top: -6px; left: 11.11111111111111%;"
          />
          <div
            class="sliderTrack"
            style="width: 11.11111111111111%; background-color: yellow;"
          />
          <div
            class="sliderStep"
          />
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: Slider should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="slider   "
  >
    <div
      class="wrapper"
    >
      <div
        class="tapArea"
      >
        <div
          class="handleWrapper"
          style="background-color: rgb(233, 233, 233);"
        >
          <div
            class="sliderHandle"
            style="left: 0%;"
          />
          <div
            class="sliderThumb"
            style="width: 28px; height: 28px; margin-left: -14px; margin-top: -14px; left: 0%; background-color: rgb(255, 255, 255);"
          />
          <div
            class="sliderTrack"
            style="width: 0%; background-color: rgb(26, 173, 25);"
          />
          <div
            class="sliderStep"
          />
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
